### PR TITLE
share: handle rows concurrently

### LINKED ISSUE #184: service/share: Implement parallelisation for `GetSharesByNamespace`
Implement parallelisation for `GetSharesByNamespace` when there are multiple rows that contain the given `namespace.ID`.

### PR DESCRIPTION
**Pull request by vgonkivs**
* Handle multiple rows concurrently;

Closes #184

### CODE REVIEW

File: service/share/share.go
Comment (Wondertan) [inline service/share/share.go:145]: Pls look at `plugin.MustCidFromNamespacedSha256`. This is a cheap operation that does not need a parallelization. In fact, new code will be even slower and use more RAM due to goroutines overhead for an operation that can be done sequentially fast enough. Instead, you should run each `ipld.GetLeavesByNamaspace` in a separate routine, as it is a blocking operation that involves a lot of blocking IO.

**Review by Wondertan — CHANGES_REQUESTED** [review]
@vgonkivs, thank you for working on this. Unfortunately, this is wrong and I'll try to explain why.

File: service/share/share.go
Comment (vgonkivs) [reply service/share/share.go:145]: Thanks for explanations, @Wondertan . Sorry, for not getting what exactly should be done from the first attempt. May I use the same approach for `ipld.GetLeavesByNamaspace`(wg+channels) ?

File: service/share/share.go
Comment (Wondertan) [inline service/share/share.go:182]: By quick look, there is one bug in here and maybe more: if `ipld.GetLeavesByNamespace` returns an error, `wg.Wait` will never unblock(line 165) and both `errCh` and `nodeCh` won't be closed.

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: service/share/share.go
Comment (vgonkivs) [reply service/share/share.go:182]: Yeah, unfortunately It has one more issue that I haven`t noticed. 
Regarding line 165 I believed that it has to be cleared by GC but it`s not. As a fallback solution I`d like to propose to call `wg.Done` also at line 180 and not return the err but storing it to local variable and return the error. So channels will be closed in this case. The minus is that we have to perform all requests to `ipld.GetLeavesByNamespace` even if error occurred.

**Comment by renaynay** [comment]
@vgonkivs Thank you very much for this PR - would you be able to take a first shot at writing a test for this?

File: service/share/share.go
Comment (liamsi) [inline service/share/share.go:182]: 🤔 Are you sure this line makes sense?

**Review by liamsi — CHANGES_REQUESTED** [review]
Please revisit this: https://github.com/celestiaorg/celestia-node/pull/241/files#r760551422

I think it would be good to add a benchmark that shows that this parallelization is justified (though it is relatively obvious that it should be faster but still).

File: service/share/share.go
Comment (vgonkivs) [reply service/share/share.go:182]: Hi @liamsi , yes , you right. The condiiton should be `if fetchErr == nil` to handle the first occurred error. But I`m thinking, do we need here to handle the first error in case of multiple errors will be occurred? Or we can return the last one and in this case we do not need this condition.

Comment (liamsi) [reply service/share/share.go:182]: Good question. I think we could (as before) return on the first caught error. There is no need to continue fetching after an error occurred.

Comment (vgonkivs) [reply service/share/share.go]: Added new context to cancel all requests if error occurred.

**Comment by vgonkivs** [comment]
Hi @renaynay , @liamsi . Thanks for the idea to add benchmarks and additional tests. I will try to provide them within next week.

**Comment by Wondertan** [comment]
> I think it would be good to add a benchmark that shows that this parallelization is justified (though it is relatively obvious that it should be faster but still).

@liamsi, the optimization is not CPU bound, i.e. we don't parallelize something calculation intensive, but we parallelize IO bound logic, so parallelization is furthermore a wrong word here, while concurrency is a right one(like it is said in PR name). Thus, we might not see any performance improvements and even performance degradation when we mock out network responses. 

Particularly, here we start traversing and requesting each row concurrently, instead of doing them in order. Imagine sending 10 network requests in order and compare this to spawning 10 concurrent goroutines. In order requires for n+1 request to wait for n response to land first, while sending concurrently allows requests to be sent independently. In the mock environment, where responses are immediate, goroutines may only add overhead, as there are no responses to wait for.

**Comment by vgonkivs** [comment]
Thanks for the explanation @Wondertan. I was a bit confused after I had received the results - the code really works slower on mock.
@liamsi , I have update PR with additional test and benchmark.

File: service/share/share.go
Comment (Wondertan) [inline service/share/share.go:170]: ?

cancel is a func that should be called. Here you just reassign the func value to nothing.

File: service/share/share.go
Comment (Wondertan) [inline service/share/share.go:150]: The code can be simplified by making an anonymous `result struct` that has both the value and the error, so only one channel of the type needs to be created and handled. That's a common pattern in such situations.

File: service/share/share.go
Comment (Wondertan) [inline service/share/share.go:153]: It's ok to have multiple defers in a func and it is a good practice to keep each defer next to the value a defer is closing/canceling/flushing or etc.

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: service/share/share.go
Comment (vgonkivs) [reply service/share/share.go]: done

File: service/share/share.go
Comment (vgonkivs) [reply service/share/share.go]: removed

File: service/share/share.go
Comment (vgonkivs) [reply service/share/share.go:155]: reworked

**Review by Wondertan — APPROVED** [review]
👍🏻

**Comment by adlerjohn** [comment]
Waiting on review from @liamsi

File: service/share/share.go
Comment (renaynay) [inline service/share/share.go:166]: Should we check for `ok` as well / make sure`result` is never nil? It could technically panic otherwise, no?

File: service/share/share_test.go
Comment (renaynay) [inline service/share/share_test.go:69]: Is there also a test to check for the correctness of the shares that are returned via this process? If not, can you please add it?

File: service/share/share.go
Comment (Wondertan) [reply service/share/share.go:166]: `resultCh` is never closed, so that is not an issue

File: service/share/share_test.go
Comment (Wondertan) [reply service/share/share_test.go:69]: WDYM by "correctness"?

File: service/share/share.go
Comment (renaynay) [reply service/share/share.go:166]: and there's no instance where result could ever be nil?

File: service/share/share_test.go
Comment (renaynay) [reply service/share/share_test.go:69]: `TestService_GetSharesByNamespace` should be modified so that if `expectedShares > 1`, then check both shares[0] and shares[1]. 

We should also expand testing around this method to cover more cases (like overflow of NID into multiple rows, not just 2)

File: service/share/share.go
Comment (Wondertan) [reply service/share/share.go:166]: Yes, you can see it from the code.

File: service/share/share_test.go
Comment (Wondertan) [reply service/share/share_test.go:69]: Should this be part of this PR?

Comment (renaynay) [reply service/share/share_test.go:69]: At least the first point, yes bc it's testing for accuracy of *all* shares returned by `GetShares`

Comment (Wondertan) [reply service/share/share_test.go:69]: @vgonkivs

Comment (vgonkivs) [reply service/share/share_test.go:71]: @renaynay, I hope I understand you correctly.
Added loop to `TestService_GetSharesByNamespace` to check all returned shares.

Comment (renaynay) [reply service/share/share_test.go:71]: Perfect.

**Review by renaynay — APPROVED** [review]
thanks so much for this @vgonkivs 

The tests are failing on a flakey test, which we have documented here: https://github.com/celestiaorg/celestia-node/issues/173

**Review by Wondertan — APPROVED** [review]

**Review by liamsi — APPROVED** [review]
Looks good! Thank you @vgonkivs 🙏🏼 

Also thanks @renaynay and @Wondertan for the reviews.

**Pull request merged**
